TDL-09-shell-EC2/deploy_instance.sh

```diff
@@ -0,0 +1,229 @@
+ 
+#  Create EC2 instance for osype with instance_type with below settings:
+print_color "green" "-------------------------Create EC2 instance---------------------------"
+# set variables
+KEY_NAME='da-key'
+VPC_NAME='da-c02-vpc'
+VPC_ID=$(aws ec2 describe-vpcs \
+    --filter "Name=tag:Name,Values=$VPC_NAME" \
+    --query Vpcs[].VpcId --output text)
+echo $VPC_ID
+PRIVATE_SUBNET_A_NAME='da-c02-private-a'
+PRIVATE_SUBNET_A_ID=$(aws ec2 describe-subnets  \
+    --filter "Name=tag:Name,Values=$PRIVATE_SUBNET_A_NAME" \
+    --query 'Subnets[].SubnetId' --output text)
+echo "${PRIVATE_SUBNET_A_ID}"
+
+SG_PRIVATE_NAME='da-c02-private-sg'
+SG_PRIVATE_ID=$(aws ec2 describe-security-groups \
+    --filters Name=vpc-id,Values="${VPC_ID}" Name=group-name,Values="${SG_PRIVATE_NAME}" \
+    --query  'SecurityGroups[*].[GroupId]'  --output text)
+echo "${SG_PRIVATE_ID}"
+
+# start_time=$(date -d "$(TZ='Australia/Melbourne' date)" +%Y%m%d%H%M%S)
+# echo ${start_time}
+## AgedTime: yyyymmddhhmm (current time + agedtime days in Australia/Melbourne Time)
+stop_time=$(date -d "$(TZ=$DEFAULT_TIME_ZONE date) $agedtime days -3600 seconds" +%Y%m%d%H%M%S)
+echo ${stop_time}
+
+# Create user_data file
+cat << EOF > ./instance_userdata.txt
+#!/bin/bash
+cat <<EOT >> /etc/motd
+'OSType: $ostype'
+"Please be notified that your instance will be terminated by  ${stop_time}"
+EOT
+EOF
+
+EC2_NAME='ostype_test'
+RULE='ec2-startstop;1900;;all'
+# Create instance
+instance_id=$(aws ec2 run-instances \
+    --image-id "$image_id" \
+    --instance-type "$instance_type" \
+    --count 1 \
+    --subnet-id "$PRIVATE_SUBNET_A_ID" \
+    --key-name  "$KEY_NAME" \
+    --security-group-ids "$SG_PRIVATE_ID" \
+    --no-associate-public-ip-address \
+    --user-data file://instance_userdata.txt  \
+    --tag-specifications "ResourceType=instance,Tags=[{Key=Name,Value=$EC2_NAME},{Key=AgedTime,Value=$stop_time},{Key=scheduler,Value=$RULE}]"  \
+    --query 'Instances[*].InstanceId' --output text)
+
+private_ip_address=$(aws ec2 describe-instances \
+    --filter "Name=tag:Name,Values=$EC2_NAME" \
+    --query "Reservations[*].Instances[*].PrivateIpAddress" --output text)
+
+#  Check instance status and wait instance is running status
+print_color "green" "------Check instance status and wait instance is running status--------"
+#instance-state-name - The state of the instance (pending | running | shutting-down | terminated | stopping | stopped ).
+while true
+do
+  state=$(aws ec2 describe-instances \
+      --instance-ids "$instance_id" \
+      --query "Reservations[*].Instances[*].State.Name" --output text)
+  echo $state
+  if [[ $state == 'running' ]]; then
+    print_color "green" "Successed. Instance is running."
+    break
+  fi
+done
+
+# Print out private ip and 
+print_color "green" "-------------------------print out instance id and private ip--------------------------"
+print_color "green" "Outputs: InstanceId=\"$instance_id\", PrivateIpAddress=\"$private_ip_address\""
```

Comment:
One more thing: clean up, delete the temp file for userdata script;)

Reply:
Ok.